Updated and polished the "Torii Endpoints" topic

yamkovoy opened the pull request:
# Changes
- Updated and polished the **Torii Endpoints** topic:
  - Added intro;
  - Streamlined the structure of the entries (operations);
  - Clarified wording;
  - Added context and cross-references, etc;
  - Added responses;
  - Added several **TODO** reminders for future updates.
- Fixed and/or removed dead links in the related topics (e.g., **Advanced** > **Metrics**).

0x009922 commented on src/reference/torii-endpoints.md:
```diff
@@ -1,69 +1,167 @@
+
+   :::
 
 ## API Version
```
It is worth mentioning that this endpoint is available if `telemetry` feature is enabled in Iroha.

It applies to other endpoints as well:

- metrics
- status

And there is also `schema` feature which enables "schema" endpoint.

### What are the features?

Those are flag to compile Iroha with different stuff inside. By default all of them are enabled. Disabling them might be beneficial for optimisation.

I think we should also document features somewhere in the reference, in a different topic.

yamkovoy replied:
Created a new issue for tracking: #465

0x009922 commented on src/reference/torii-endpoints.md:
````diff
@@ -1,69 +1,167 @@
+The first HTTP request to this endpoint requires a standard set of `WebSocket` headers.
 
-Via this endpoint, the client first provides the starting block number (i.e. height) in the subscription request. After
-sending the confirmation message, the server starts streaming all the blocks from the given block number up to the
-current block and continues to stream blocks as they are added to the blockchain.
+**Example**:
+
+```http
+Host: example.com:8000
+Upgrade: websocket
+Connection: Upgrade
+Sec-WebSocket-Key: x3JJHMbDL1EzLkh9GBhXDw==
+Sec-WebSocket-Version: 13
+Sec-WebSocket-Extensions: permessage-deflate, client_max_window_bits
+```
````
I am not sure we should cover details of how WebSocket works. It is pretty much a standard and API users might use any convenient WebSocket client they want. Such stuff about headers is performed under the hood and doesn't really matter for developers.

What matters and what is related to Iroha, however, is the flow of handshake and data exchange.

0x009922 commented on src/reference/torii-endpoints.md:
```diff
@@ -1,69 +1,167 @@
+
+#### Responses
+
+| Code | Response              | Description                                                                   |
+| :--: | --------------------- | ----------------------------------------------------------------------------- |
+| 101  | Switching Protocols   | The protocol is successfully switched from `HTTP` to `WebSocket`.             |
+| 400  | Bad Request           | The server will not process the request due to an issue on the client side.   |
+| 401  | Unauthorized          | The client lacks valid credentials for the request.                           |
+| 500  | Internal Server Error | The server encountered an unexpected issue and could not fulfill the request. |
```
- 101 - might occur, doesn't matter for Torii docs
- 400 - never
- 401 - never, this request doesn't require authorization
- 500 - might occur for any request, isn't worth to mention

I would say, the whole table is not needed, as well as "responses" section.

0x009922 commented on src/reference/torii-endpoints.md:
```diff
@@ -1,69 +1,167 @@
+| Code | Response              | Description                                                                   |
+| :--: | --------------------- | ----------------------------------------------------------------------------- |
+| 200  | OK                    | Returns a subset of configuration parameters serialized into JSON format.     |
+| 401  | Unauthorized          | The client lacks valid credentials for the request.                           |
+| 500  | Internal Server Error | The server encountered an unexpected issue and could not fulfill the request. |
```
401 never happens; 500 is needless to mention.

0x009922 commented on src/reference/torii-endpoints.md:
````diff
@@ -1,69 +1,167 @@
+
+**Example**:
 
 ```json
+200 OK:
 {
   "logger": {
     "level": "TRACE"
   }
 }
 ```
````
````suggestion
```json
{
  "logger": {
    "level": "TRACE"
  }
}
```
````

0x009922 commented on src/reference/torii-endpoints.md:
```diff
@@ -73,59 +171,111 @@ For possible values please refer to the configuration reference (TODO:
+| Code | Response              | Description                                                                     |
+| :--: | --------------------- | ------------------------------------------------------------------------------- |
+| 202  | Accepted              | The request to update the configuration is accepted and is due to be processed. |
+| 401  | Unauthorized          | The client lacks valid credentials for the request.                             |
+| 500  | Internal Server Error | The server encountered an unexpected issue and could not fulfill the request.   |
```
401 never happens; 500 is needless to mention.

0x009922 commented on src/reference/torii-endpoints.md:
```diff
@@ -73,59 +171,111 @@ For possible values please refer to the configuration reference (TODO:
-[`EventSubscriptionRequest`](/reference/data-model-schema#eventsubscriptionrequest). Then the server sends an
-[`EventMessage`](/reference/data-model-schema#eventmessage). Messages are SCALE-encoded[^1].
+- **Protocol**: `HTTP` upgraded to `WebSocket`
+- **Encoding**: `SCALE`
```
I think leaving a footnote for `SCALE` makes sense.

0x009922 commented on src/reference/torii-endpoints.md:
```diff
@@ -73,59 +171,111 @@ For possible values please refer to the configuration reference (TODO:
 
 :::
 
 ## Events
```
What I said in the block stream section about WebSocket applies here as well.

0x009922 commented on src/reference/torii-endpoints.md:
```diff
@@ -73,59 +171,111 @@ For possible values please refer to the configuration reference (TODO:
+| Code | Response              | Description                                                                   |
+| :--: | --------------------- | ----------------------------------------------------------------------------- |
+| 200  | Health Status         | Returns the current status of the peer submitting the request.                |
+| 401  | Unauthorized          | The client lacks valid credentials for the request.                           |
+| 500  | Internal Server Error | The server encountered an unexpected issue and could not fulfill the request. |
```
401 never happens; 500 is needless to mention.

0x009922 commented on src/reference/torii-endpoints.md:
```diff
@@ -73,59 +171,111 @@ For possible values please refer to the configuration reference (TODO:
+| Code | Response              | Description                                                                   |
+| :--: | --------------------- | ----------------------------------------------------------------------------- |
+| 200  | Metrics               | Returns a report on 8 out of 10 Prometheus metrics.                           |
+| 401  | Unauthorized          | The client lacks valid credentials for the request.                           |
+| 500  | Internal Server Error | The server encountered an unexpected issue and could not fulfill the request. |
```
401 never happens; 500 is needless to mention.

0x009922 commented on src/reference/torii-endpoints.md:
```diff
@@ -226,15 +385,35 @@ Whether each prerequisite object was found and [`FindError`](/reference/data-mod
+
+#### Responses
+
+| Code | Response              | Description                                                                   |
+| :--: | --------------------- | ----------------------------------------------------------------------------- |
+| 200  | Iroha Status          | Returns the Iroha network status report.                                      |
+| 401  | Unauthorized          | The client lacks valid credentials for the request.                           |
+| 500  | Internal Server Error | The server encountered an unexpected issue and could not fulfill the request. |
+
+The `200 Iroha Status` response is generated in accordance with the encoding specified in the header of the request.
```
```suggestion
| Code | Response              | Description                                                                   |
| :--: | --------------------- | ----------------------------------------------------------------------------- |
| 200  | Iroha Status          | Returns the Iroha network status report. The response is generated in accordance with the encoding specified in the header of the request and is specified in `Content-Type` response header.  |
```

0x009922 commented on src/reference/torii-endpoints.md:
````diff
@@ -303,13 +466,48 @@ These samples represent the same data:
 
 ::: code-group
 
+```json [/status]
+struct Status {
+  "peers": 4,
+  "blocks": 5,
+  "txs_accepted": 31,
+  "txs_rejected": 3,
+  "uptime": {
+    "secs": 5,
+    "nanos": 937000000
+  },
+  "view_changes": 2,
+  "queue_size": 18
+}
+
+struct Uptime {
+    secs: 5,
+    nanos: 937000000
+}
+```
````
Please use actual JSON as it used to be before.

0x009922 commented on src/reference/torii-endpoints.md:
```diff
@@ -329,23 +527,22 @@ returns the corresponding JSON value.
+| 200  | Transaction Accepted                     | Transaction has been accepted, but is not yet guaranteed to have passed consensus. |
+| 400  | Transaction Rejected (Malformed)         | Transaction is rejected due to being malformed.                                    |
+| 401  | Transaction Rejected (Improperly signed) | Transaction is rejected due to being improperly signed.                            |
+| 500  | Internal Server Error                    | The server encountered an unexpected issue and could not fulfill the request.      |
```
500 is needless to mention

0x009922 reviewed (changes requested):
LGTM!

Torii endpoints now seem to be more structured and detailed. For me, as a developer, it became clearer.

However, there are a lot of inaccurate details as well.

yamkovoy commented:
# Changes
- Addressed comments by @0x009922;
- Added requested **TODO** reminders;
- Added several notes on Iroha features/flags;
- Added **Schema** and **Pending Transactions** operations;
- Other minor changes.

# Notes & Questions
- In the `guide/advanced/metrics.md` topic, the **How to use metrics** subtopic still has the instructions. I couldn't leave it empty, nor could I roll it back to the previous version, since the config file mentioned in it no longer exists.
  > @0x009922, any advice on what information should we place there until PR #397 is ready to be merged?
- In both `guide/advanced/metrics.md` and `reference/torii-endpoints.md` topics, @0x009922 pointed out in his comments some inaccurate bits of information on how to configure metrics and Torii API. Since I gathered the instructions that I placed there from the exisiting **Configure Iroha** topics (i.e., [Peer Configuration](https://docs.iroha.tech/guide/configure/peer-configuration.html), [Client Configuration](https://docs.iroha.tech/guide/configure/client-configuration.html), etc.), I am struggling to understand what exactly was inaccurate. I expected there to be inconsistencies that I was hoping could be resolved within this PR, however, as I understand your comments, the added instructions were completely off the mark.
  > @0x009922, is it some misunderstanding from my part or have there been some significant changes that are yet to be documented?
  > Please, do not take this question as offensive, since I'm merely trying to understand my mistake and help us achieve the best possible result that can only be found in cooperation.

0x009922 commented:
@yamkovoy, I am totally fine with your questions, and I can see that my comments might look unclear.

The configuration sections you have provided are inaccurate primarily because the whole config system is currently under overhaul by me. I do not provide you any information yet about how to document it properly, and ask you just to leave TODO marks so that I can pick it up later and fill the gaps.

Thus, it would be totally fine in scope of this PR to leave TODO marks instead of trying to cover configuration topics, IMO.

If it would mean to leave the whole Metrics topic empty - so be it.

Please wait until I prepare the configuration reference which will be the foundation in future. (#397)

0x009922 commented on src/reference/torii-endpoints.md:
```diff
@@ -1,46 +1,127 @@
-- **Endpoint:** `/api_version`
+::: info
+
+This operation requires the Iroha 2 network to be established with the `telemetry` feature enabled.
```
```suggestion
This operation requires the particular Iroha node you are making request to to be compiled with the `telemetry` feature enabled.
```

0x009922 commented on src/reference/torii-endpoints.md:
```diff
@@ -1,46 +1,127 @@
+
+#### Data Exchange
+
+After a successful handshake, the client must send a [`BlockSubscriptionRequest`](/reference/data-model-schema#blocksubscriptionrequest) request with the starting block number provided (i.e., the `height` value). Then, upon sending the confirmation and [`BlockMessage`](/reference/data-model-schema#blockmessage) messages, the server starts streaming all of the blocks, beginning with the block spec ...
```
```suggestion
After establishing a WebSocket connection, the client must send a [`BlockSubscriptionRequest`](/reference/data-model-schema#blocksubscriptionrequest) request with the starting block number provided (i.e., the `height` value). Then, upon sending the confirmation and [`BlockMessage`](/reference/data-model-schema#blockmessage) messages, the server starts streaming all of the blocks, beginning with the block specified with `height` up to the most recent one, and then continues to stream new blocks as they are added to the blockchain.
```

0x009922 commented on src/reference/torii-endpoints.md:
````diff
@@ -172,37 +313,77 @@ view_changes 0
+| :--: | -------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| 200  | OK       | Returns a list of pending transactions as [`SignedTransaction`](data-model-schema.md#signedtransaction) objects encoded with `SCALE`; must be decoded by the client. |
+
+**Example**:
+
+```[SCALE]
+[
+  34,
+  86,
+  97,
+  108,
+  117,
+  101,
+  34
+]
+```
````
Providing an example here is not useful. I would omit it entirely.

**Note:** providing SCALE example for the Status endpoint still makes sense.

0x009922 reviewed (approved):
LGTM.

High attention to detail. In some places even a little bit too much (imo), but it's fine.
